File: c/library_2/determine_opposite_halfedge_quad.c

#ifndef _INCLUDE_BASIC_
#define _INCLUDE_BASIC_

#include "basic_emxArray_fncs.c"

#endif

/*
 * DETERMINE_OPPOSITE_HALFEDGE_QUAD Determine the opposite half-edge.
 *  DETERMINE_OPPOSITE_HALFEDGE_QUAD(NV,QUADS,OPPHES) Determines
 *  the opposite half-edge for a quadrilateral or mixed (quad-dominant) mesh.
 *  The following explains the input and output arguments.
 *
 *  OPPHES = DETERMINE_OPPOSITE_HALFEDGE_QUAD(NV,QUADS)
 *  OPPHES = DETERMINE_OPPOSITE_HALFEDGE_QUAD(NV,QUADS,OPPHES)
 *  Computes mapping from each half-edge to its opposite half-edge.
 *
 *  Convention: Each half-edge is indicated by <face_id,local_edge_id>.
 *  We assign 2 bits to local_edge_id (starts from 0).
 *
 *  See also DETERMINE_OPPOSITE_HALFEDGE
 */
void determine_opposite_halfedge_quad(int32_T nv, const emxArray_int32_T *quads,
  emxArray_int32_T *opphes)
{
  emxArray_int32_T *is_index;
  int32_T nquads;
  int32_T i0;
  int32_T ii;
  boolean_T exitg1;
  int32_T b_is_index[3];
  int32_T c_is_index[4];
  emxArray_int32_T *v2nv;
  emxArray_int32_T *v2he;
  int32_T ne;
  static const int8_T iv0[3] = { 1, 2, 0 };

  static const int8_T iv1[4] = { 1, 2, 3, 0 };

  int32_T quads_idx_0;
  int32_T i1;
  static const int8_T iv2[4] = { 2, 3, 4, 1 };

  static const int8_T iv3[3] = { 2, 3, 1 };

  emxInit_int32_T(&is_index, 1);

  /*  Number of edges per element */
  nquads = quads->size[0];

  /* % First, build is_index to store starting position for each vertex. */
  i0 = is_index->size[0];
  is_index->size[0] = nv + 1;
  emxEnsureCapacity((emxArray__common *)is_index, i0, (int32_T)sizeof(int32_T));
  for (i0 = 0; i0 <= nv; i0++) {
    is_index->data[i0] = 0;
  }

  ii = 0;
  exitg1 = FALSE;
  while ((exitg1 == 0U) && (ii + 1 <= nquads)) {
    if (quads->data[ii] == 0) {
      nquads = ii;
      exitg1 = TRUE;
    } else {
      if (quads->data[ii + quads->size[0] * 3] == 0) {
        for (i0 = 0; i0 < 3; i0++) {
          b_is_index[i0] = is_index->data[quads->data[ii + quads->size[0] * i0]]
            + 1;
        }

        for (i0 = 0; i0 < 3; i0++) {
          is_index->data[quads->data[ii + quads->size[0] * i0]] = b_is_index[i0];
        }
      } else {
        for (i0 = 0; i0 < 4; i0++) {
          c_is_index[i0] = is_index->data[quads->data[ii + quads->size[0] * i0]]
            + 1;
        }

        for (i0 = 0; i0 < 4; i0++) {
          is_index->data[quads->data[ii + quads->size[0] * i0]] = c_is_index[i0];
        }
      }

      ii++;
    }
  }

  is_index->data[0] = 1;
  for (ii = 1; ii <= nv; ii++) {
    is_index->data[ii] += is_index->data[ii - 1];
  }

  emxInit_int32_T(&v2nv, 1);
  emxInit_int32_T(&v2he, 1);
  ne = nquads << 2;
  i0 = v2nv->size[0];
  v2nv->size[0] = ne;
  emxEnsureCapacity((emxArray__common *)v2nv, i0, (int32_T)sizeof(int32_T));

  /*  Vertex to next vertex in each halfedge. */
  i0 = v2he->size[0];
  v2he->size[0] = ne;
  emxEnsureCapacity((emxArray__common *)v2he, i0, (int32_T)sizeof(int32_T));

  /*  Vertex to half-edge. */
  for (ii = 0; ii + 1 <= nquads; ii++) {
    if (quads->data[ii + quads->size[0] * 3] == 0) {
      for (i0 = 0; i0 < 3; i0++) {
        v2nv->data[is_index->data[quads->data[ii + quads->size[0] * i0] - 1] - 1]
          = quads->data[ii + quads->size[0] * iv0[i0]];
      }

      ne = (ii + 1) << 2;
      for (i0 = 0; i0 < 3; i0++) {
        v2he->data[is_index->data[quads->data[ii + quads->size[0] * i0] - 1] - 1]
          = i0 + ne;
      }

      for (i0 = 0; i0 < 3; i0++) {
        b_is_index[i0] = is_index->data[quads->data[ii + quads->size[0] * i0] -
          1] + 1;
      }

      for (i0 = 0; i0 < 3; i0++) {
        is_index->data[quads->data[ii + quads->size[0] * i0] - 1] =
          b_is_index[i0];
      }
    } else {
      for (i0 = 0; i0 < 4; i0++) {
        v2nv->data[is_index->data[quads->data[ii + quads->size[0] * i0] - 1] - 1]
          = quads->data[ii + quads->size[0] * iv1[i0]];
      }

      ne = (ii + 1) << 2;
      for (i0 = 0; i0 < 4; i0++) {
        v2he->data[is_index->data[quads->data[ii + quads->size[0] * i0] - 1] - 1]
          = i0 + ne;
      }

      for (i0 = 0; i0 < 4; i0++) {
        c_is_index[i0] = is_index->data[quads->data[ii + quads->size[0] * i0] -
          1] + 1;
      }

      for (i0 = 0; i0 < 4; i0++) {
        is_index->data[quads->data[ii + quads->size[0] * i0] - 1] =
          c_is_index[i0];
      }
    }
  }

  for (ii = nv - 1; ii > 0; ii--) {
    is_index->data[ii] = is_index->data[ii - 1];
  }

  is_index->data[0] = 1;

  /* % Set opphes */
  if (opphes->size[0] == 0) {
    quads_idx_0 = quads->size[0];
    i0 = opphes->size[0] * opphes->size[1];
    opphes->size[0] = quads_idx_0;
    opphes->size[1] = 4;
    emxEnsureCapacity((emxArray__common *)opphes, i0, (int32_T)sizeof(int32_T));
    for (i0 = 0; i0 < 4; i0++) {
      ne = quads_idx_0 - 1;
      for (i1 = 0; i1 <= ne; i1++) {
        opphes->data[i1 + opphes->size[0] * i0] = 0;
      }
    }
  } else {
    i0 = opphes->size[0] * opphes->size[1];
    opphes->size[0] = opphes->size[0];
    opphes->size[1] = 4;
    emxEnsureCapacity((emxArray__common *)opphes, i0, (int32_T)sizeof(int32_T));
    for (i0 = 0; i0 < 4; i0++) {
      ne = opphes->size[0] - 1;
      for (i1 = 0; i1 <= ne; i1++) {
        opphes->data[i1 + opphes->size[0] * i0] = 0;
      }
    }
  }

  for (ii = 0; ii + 1 <= nquads; ii++) {
    i0 = (quads->data[ii + quads->size[0] * 3] != 0) + 3;
    for (quads_idx_0 = 0; quads_idx_0 + 1 <= i0; quads_idx_0++) {
      if (opphes->data[ii + opphes->size[0] * quads_idx_0] != 0) {
      } else {
        if (quads->data[ii + quads->size[0] * 3] != 0) {
          ne = quads->data[ii + quads->size[0] * (iv2[quads_idx_0] - 1)];
        } else {
          ne = quads->data[ii + quads->size[0] * (iv3[quads_idx_0] - 1)];
        }

        /*  LOCATE: Locate index col in v2nv(first:last) */
        i1 = is_index->data[ne] - 1;
        for (ne = is_index->data[ne - 1] - 1; ne + 1 <= i1; ne++) {
          if (v2nv->data[ne] == quads->data[ii + quads->size[0] * quads_idx_0])
          {
            opphes->data[ii + opphes->size[0] * quads_idx_0] = v2he->data[ne];

            /*  opphes(heid2fid(opp),heid2leid(opp)) = ii*4+jj-1; */
            opphes->data[((int32_T)((uint32_T)v2he->data[ne] >> 2U) +
                          opphes->size[0] * (v2he->data[ne] - ((v2he->data[ne] >>
              2) << 2))) - 1] = ((ii + 1) << 2) + quads_idx_0;
          }
        }

        /*  Check for consistency */
      }
    }
  }

  emxFree_int32_T(&v2he);
  emxFree_int32_T(&v2nv);
  emxFree_int32_T(&is_index);
}

void determine_opposite_halfedge_quad_initialize(void)
{
}

void determine_opposite_halfedge_quad_terminate(void)
{
  /* (no terminate code required) */
}

File: matlab/hi/2d/determine_opposite_halfedge_quad.m
function sibhes = determine_opposite_halfedge_quad(nv, quads, sibhes) %#codegen 
%DETERMINE_OPPOSITE_HALFEDGE_QUAD Determine the opposite half-edge.
% DETERMINE_OPPOSITE_HALFEDGE_QUAD(NV,QUADS,SIBHES) Determines
% the opposite half-edge for a quadrilateral or mixed (quad-dominant) mesh.
% The following explains the input and output arguments.
%
% SIBHES = DETERMINE_OPPOSITE_HALFEDGE_QUAD(NV,QUADS)
% SIBHES = DETERMINE_OPPOSITE_HALFEDGE_QUAD(NV,QUADS,SIBHES)
% Computes mapping from each half-edge to its opposite half-edge.
%
% Convention: Each half-edge is indicated by <face_id,local_edge_id>.
% We assign 2 bits to local_edge_id (starts from 0).
%
% See also DETERMINE_OPPOSITE_HALFEDGE

nepE = int32(4);  % Number of edges per element
next3 = int32([2,3,1]);
next4 = int32([2,3,4,1]);
inds3 = int32(1:3);
inds4 = int32(1:4);

nquads = int32(size(quads,1));

%% First, build is_index to store starting position for each vertex.
is_index = zeros(nv+1,1, 'int32');
for ii=1:nquads
    if quads(ii,1)==0; nquads=ii-1; break; end
    if quads(ii,4)==0
        is_index(quads(ii,inds3)+1) = is_index(quads(ii,inds3)+1) + 1;
    else
        is_index(quads(ii,inds4)+1) = is_index(quads(ii,inds4)+1) + 1;
    end
end
is_index(1) = int32(1);
for ii=1:nv
    is_index(ii+1) = is_index(ii) + is_index(ii+1);
end

ne = nquads*nepE;
v2nv = nullcopy(zeros( ne,1, 'int32'));  % Vertex to next vertex in each halfedge.
v2he = nullcopy(zeros( ne,1, 'int32'));  % Vertex to half-edge.
for ii=1:nquads
    if quads(ii,4)==0
        v2nv(is_index( quads(ii,inds3))) = quads(ii,next3);
        v2he(is_index( quads(ii,inds3))) = 4*ii-1+inds3;
        is_index(quads(ii,inds3)) = is_index(quads(ii,inds3)) + 1;
    else
        v2nv(is_index( quads(ii,inds4))) = quads(ii,next4);
        v2he(is_index( quads(ii,inds4))) = 4*ii-1+inds4;
        is_index(quads(ii,inds4)) = is_index(quads(ii,inds4)) + 1;
    end
end
for ii=nv-1:-1:1; is_index(ii+1) = is_index(ii); end
is_index(1)=1;

%% Set sibhes
if nargin<3 || isempty(sibhes)
    sibhes = zeros(size(quads,1), nepE, 'int32');
else
    assert( size(sibhes,1)>=nquads && size(sibhes,2)>=nepE);
    sibhes(:) = 0;
end

for ii=1:nquads
    
    for jj=1:3+int32(quads(ii,4)~=0)
        if sibhes(ii,jj); continue; end
        v = quads(ii,jj); 
        if quads(ii,4)~=0; 
            vn = quads(ii,next4(jj));
        else
            vn = quads(ii,next3(jj));
        end

        % LOCATE: Locate index col in v2nv(first:last)
        found = 0;
        for index = is_index(vn):is_index(vn+1)-1
            if v2nv(index)==v
                opp = v2he(index);
                sibhes(ii,jj) = opp;
                % sibhes(heid2fid(opp),heid2leid(opp)) = ii*4+jj-1;
                sibhes( bitshift( uint32(opp),-2),mod(opp,4)+1) = ii*4+jj-1;

                found = found + 1;
            end
        end

        % Check for consistency
        if found>1
            error( 'Input mesh is not an oriented manifold.');
        elseif ~found
            for index = is_index(v):is_index(v+1)-1
                if v2nv(index)==vn && int32(bitshift(uint32(v2he(index)),-2))~=ii
                    if nargin==3
                        error( 'Input mesh is not oriented.');
                    else
                        sibhes = zeros(0,nepE, 'int32'); return;
                    end
                end
            end
        end
    end
end
